File: ledger-models-python/fintekkers/wrappers/models/security.py
from fintekkers.models.security.identifier.identifier_pb2 import IdentifierProto
from fintekkers.models.security.security_pb2 import SecurityProto

from fintekkers.models.position.field_pb2 import *
from fintekkers.models.position.measure_pb2 import MeasureProto

from uuid import UUID
from datetime import datetime
from fintekkers.wrappers.models.security_identifier import Identifier

from fintekkers.wrappers.models.util.fintekkers_uuid import FintekkersUuid
from fintekkers.wrappers.models.util.serialization import ProtoSerializationUtil

class IFinancialModelObject:
    def get_field(field:FieldProto) -> object:
        pass

    def get_measure(measure:MeasureProto) -> object:
        pass

    def get_fields() -> set[FieldProto]:
        pass

    def get_measures() -> set[MeasureProto]:
        pass

    def get_as_of() -> datetime:
        pass

class RawDataModelObject:
    def __init__(self, id: UUID, as_of: datetime):
        self.id = id
        self.as_of = as_of

class Security():
    def __init__(self, proto:SecurityProto):
        self.proto:SecurityProto = proto

    def __str__(self) -> str:
        return f"ID[{str(self.get_id())}], {self.get_security_id()}[{self.proto.issuer_name}]"

    def get_fields(self) -> list[FieldProto]:
        return [
            ID, SECURITY_ID, AS_OF, ASSET_CLASS, IDENTIFIER
        ]

    def get_field(self, field:FieldProto) -> object:
        if field in (ID, SECURITY_ID):
            return self.get_id()
        elif field == AS_OF:
            return self.get_as_of()
        elif field == ASSET_CLASS:
            return self.get_asset_class()
        elif field == PRODUCT_CLASS:
            return self.get_product_class()
        elif field == PRODUCT_TYPE:
            return self.get_product_type()
        elif field == IDENTIFIER:
            return self.get_security_id()
        elif field in (TENOR, ADJUSTED_TENOR):
            raise ValueError("Not implemented yet")
        elif field == MATURITY_DATE:
            raise ValueError("Not implemented yet")
        else:
            raise ValueError(f"Field not mapped in Security wrapper: {field}")

    def get_id(self) -> UUID:
        uuid:FintekkersUuid = ProtoSerializationUtil.deserialize(self.proto.uuid)
        return uuid.uuid
    
    def get_as_of(self) -> datetime:
        as_of:datetime = ProtoSerializationUtil.deserialize(self.proto.as_of)
        return as_of
        
    def get_asset_class(self) -> str:
        return self.proto.asset_class
    
    def get_product_class(self) -> str:
        raise ValueError("Not implemented yet. See Java implementation for reference")
    
    def get_product_type(self) -> object:
        raise ValueError("Not implemented yet. See Java implementation for reference")
    
    def get_security_id(self) -> Identifier:
        id:IdentifierProto = self.proto.identifier
        return Identifier(id)

    def __str__(self):
        return f'ID[{str(self.get_security_id())}], {type(self).__name__}[{self.proto.issuer_name}]'

    def __eq__(self, other):
        if isinstance(other, Security):
            return self.get_id() == other.get_id()
        else:
            return False

    def __lt__(self, other):
        if isinstance(other, Security):
            return self.get_id() < other.get_id()
        else:
            return False

    def __hash__(self):
        return hash(self.get_id())

# class Security(RawDataModelObject, IFinancialModelObject):
#     def __init__(self, id: uuid.UUID, issuer: str, as_of: str, settlement_currency):
#         super().__init__(id, as_of)
#         self.issuer = issuer
#         self.settlement_currency = settlement_currency
#         self.identifier = None
#         self.product_type = None
#         self.description = None

#     def get_settlement_currency(self):
#         return self.settlement_currency

#     def is_cash(self):
#         return False

#     def get_issuer(self):
#         return self.issuer

#     def get_asset_class(self):
#         return 'Unclassified'

#     def get_quantity_type(self) -> SecurityQuantityTypeProto:
#         return SecurityQuantityTypeProto.UNKNOWN_QUANTITY_TYPE

#     def get_security_id(self):
#         return self.identifier

#     def set_security_id(self, identifier):
#         self.identifier = identifier

#     def get_product_class(self):
#         return type(self).__name__

#     def get_product_type(self):
#         return SecurityTypeProto.SecurityTypeProto

#     def set_product_type(self, product_type):
#         self.product_type = product_type

#     def get_fields(self) -> Set[FieldProto]:
#         return {FieldProto.ID, FieldProto.ASSET_CLASS, FieldProto.PRODUCT_CLASS}

#     def get_measure(self, measure: MeasureProto):
#         raise NotImplementedError

#     def get_measures(self) -> Set[MeasureProto]:
#         raise NotImplementedError

#     def get_security_type(self):
#         raise RuntimeError('Not supported. Need to code this in')

#     def get_display_description(self):
#         if self.description is not None:
#             return self.description
#         elif self.identifier is not None:
#             return str(self.identifier)
#         else:
#             return str(self)

#     def get_description(self):
#         return self.description

#     def set_description(self, description):
#         self.description = description



# class CashSecurity(Security):
#     pass

File: ledger-models-java/src/main/java/common/models/security/Security.java
package common.models.security;

import common.models.IFinancialModelObject;
import common.models.RawDataModelObject;
import common.models.postion.Field;
import common.models.postion.Measure;
import common.models.security.identifier.Identifier;
import fintekkers.models.security.SecurityTypeProto;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.*;

/***
 * Generally shouldn't be used except for tests, or absolute emergencies.
 */
public class Security extends RawDataModelObject implements Comparable, IFinancialModelObject {
    private final String issuer;
    private final CashSecurity settlementCurrency;
    protected Identifier identifier;

    private String productType;

    private String description;

    public Security(UUID id, String issuer, ZonedDateTime asOf, CashSecurity settlementCurrency) {
        super(id, asOf);
        this.issuer = issuer;
        this.settlementCurrency = settlementCurrency;
    }

    public CashSecurity getSettlementCurrency() {
        return settlementCurrency;
    }

    public boolean isCash() {
        return false;
    }

    public String getIssuer() {
        return this.issuer;
    }

    /**
     * Returns a high-level asset class for a security, such as 'Equity', 'FixedIncome', etc.
     *
     * @return Returns unclassified for basic securities, and a suitable value for other security values.
     */
    public String getAssetClass() {
        return "Unclassified";
    }

    public QuantityType getQuantityType() {
        return QuantityType.UNITS;
    }

    public Identifier getSecurityId() {
        return identifier;
    }

    public void setSecurityId(Identifier identifier) {
        this.identifier = identifier;
    }


    /***
     * Plumbing for positions
     */
    public Object getField(Field field) {
        return switch (field) {
            case ID, SECURITY_ID -> getID();
            case AS_OF -> getAsOf();
            case ASSET_CLASS -> getAssetClass();
            case PRODUCT_CLASS -> getProductClass();
            case PRODUCT_TYPE -> getProductType();
            case IDENTIFIER -> getSecurityId();
            case TENOR, ADJUSTED_TENOR -> Tenor.UNKNOWN_TENOR;
            case MATURITY_DATE -> LocalDate.of(2399, 12, 31);
            default -> throw new RuntimeException(String.format("Field not found %s", field));
        };
    }

    @Override
    public BigDecimal getMeasure(Measure measure) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Set<Measure> getMeasures() {
        throw new UnsupportedOperationException();
    }

    /**
     * The type of security as modeled in code layers. Examples: BondSecurity, EquitySecurity, etc.
     * @return String representation of the class, generally matches the class name in Java.
     */
    private String getProductClass() {
        return getClass().getSimpleName();
    }

    /**
     * The type of security that is suitable for user display. This will be a lower level of
     * <p>
     * BLAH. Maybe need to code in a product hierarchy now? TODO TODO TODO
     * <p>
     * 1/ Create a product type enum with opinionated values
     * 2/ Create a generic value to let people override
     * <p>
     * Needs an overridable concept. (a) How would that work; (b) What would the performance penalty be?
     *
     * @return TODO NEED TO DO THIS
     */
    public ProductType getProductType() {
        return ProductType.UNCLASSIFIED;
    }

    public void setProductType(String productType) {
        this.productType = productType;
    }

    public Set<Field> getFields() {
        return new HashSet<>(Arrays.asList(Field.ID, Field.ASSET_CLASS, Field.PRODUCT_CLASS));
    }

    @Override
    public String toString() {
        return String.format("ID[%s], %s[%s]", getID().toString(), getClass().getSimpleName(), getIssuer());
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof Security) {
            return Objects.equals(((Security) obj).getID(), getID());
        } else {
            return false;
        }
    }

    @Override
    public int compareTo( Object obj) {
        if(obj instanceof Security) {
            return getID().compareTo(((Security)obj).getID());
        }

        return -1;
    }

    @Override
    public int hashCode() {
        return getID().hashCode();
    }

    public SecurityTypeProto getSecurityType() {
        throw new RuntimeException("Not supported. Need to code this in");
    }

    /**
     * @return If an explicit description is set then it is return, otherwise a generic description is returned.
     * The description is subject to change and should NEVER be parsed. The goal is for this to be human-readable.
     */
    public String getDisplayDescription() {
        return description != null ? description :
                identifier != null ? identifier.toString() :
                toString();
    }

    /**
     * @return the explicitly set description if set
     */
    public String getDescription() {
        return description;
    }

    /**
     * @param description Human-readable description
     */
    public void setDescription(String description) {
        this.description = description;
    }
}
